### PR TITLE
python debugging module - closes #671

### LINKED ISSUE #671: python debugging module

### PR DESCRIPTION
**Pull request by BenHesketh21**

### CODE REVIEW

File: topics/python/modules/debugging/README.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+We typically work with both static analysis and dynamic analysis.
+
+## Static analysis
+
+Static analysis is a method of debugging done by examining the code without executing it.
+To do this, you need to be able to understand the structure of the code, normally best to do for your own code.
+This kind of analysis helps to ensure the code adheres to industry standards.
+This can be done manually or using tools.
+
+Static analysis can reveal errors that would not have revealed themselves until a diaster occurs weeks, months or years later.
```
Comment (LukeBenson) [inline topics/python/modules/debugging/README.md]:
```suggestion
Static analysis can reveal errors that may not have revealed themselves until a disaster occurs weeks, months or years later.
```

File: topics/python/modules/debugging/README.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+The main problem with this kind of debugging is that it finds bugs but does not prevent them.
+
+## Debugging tools
+
+There are many tools available for debugging, but there is also one built into Python.
+This debugging tool is called `pdb`, **P**ython **D**e**B**ugging.
+This tool allows you to go through code line by line to identify errors.
+
+`pdb.set_trace()` allows you to set a point you wish to start debugging from.
+
```
Comment (LukeBenson) [inline topics/python/modules/debugging/README.md]: Think you need a high level description of what pdb commands do (p,n,q and how you run pdb)

File: topics/python/modules/debugging/README.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Now when you run the program it will show the line, `a = "aaa"`. If you try entering `p a`, you will get an error as you have to pass the line for the variable to be defined so type in `n` and press enter. Then try `p a` again, you should get a return of 'aaa'.
+
+Now if you type `q` and press enter you will quit the program. 
+
+Now run the program again go the final line by typing `n` and enter repeatedly. Now type `p final` and you should get `aaabbbccc` returned.
+
+## Exercises
+
+Debug the following programs. For the first 2 perform static analysis. For the 3rd and 4th of the exercises use dynamic analysis either manually or through `pdb`.
```
Comment (LukeBenson) [inline topics/python/modules/debugging/README.md]:
```suggestion
Debug the following programs. For the first 2 perform static analysis. For the 3rd and 4th exercises use dynamic analysis either manually or through `pdb`.
```

File: topics/python/modules/debugging/README.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+Static analysis can reveal errors that would not have revealed themselves until a diaster occurs weeks, months or years later.
+
+## Dynamic analysis
+
+Dynamic analysis is the testing and evaluation of a program by executing data in real time.
+
+The objective of this kind of analysis is to find errors whilst the program is running, rather than examining code offline.
+This can also be done manually or using tools.
+
+Whilst analyzing dynamically you can debug a program in all scenarios for which it was designed.
```
Comment (BenHesketh21) [inline topics/python/modules/debugging/README.md]:
```suggestion
Whilst analysing dynamically you can debug a program in all scenarios for which it was designed.
```

File: topics/python/modules/debugging/README.md (hunk 2 above)
Comment (BenHesketh21) [inline topics/python/modules/debugging/README.md]:
```suggestion
When you run the code with this line inserted into it, the console will come up with the first line of the code. You can navigate through your code using the following:
- `n` - Next line
- `q` - Quit
- `p` - Reference variables 
```

**Pull request merged**